builtin/credential/aws/path_login.go

```diff
@@ -1317,15 +1330,27 @@ func (b *backend) pathLoginUpdateIam(ctx context.Context, req *logical.Request,
 		case strutil.StrListContains(roleEntry.BoundIamPrincipalIDs, callerUniqueId): // check 1 passed
 		case !roleEntry.ResolveAWSUniqueIDs && strutil.StrListContains(roleEntry.BoundIamPrincipalARNs, entity.canonicalArn()): // check 2 passed
 		default:
-			// evaluate check 3
+			// evaluate check 3 -- only try to look up full ARNs if there's a wildcard ARN in BoundIamPrincipalIDs.
+			hasWildcardBind := false
+			for _, principalARN := range roleEntry.BoundIamPrincipalARNs {
+				if strings.HasSuffix(principalARN, "*") {
+					hasWildcardBind = true
+					break
+				}
+			}
+
+			if !hasWildcardBind {
+				return logical.ErrorResponse(fmt.Sprintf("IAM Principal %q does not belong to the role %q", callerID.Arn, roleName)), nil
```

Comment:
```suggestion
				return logical.ErrorResponse("IAM Principal %q does not belong to the role %q", callerID.Arn, roleName), nil
```

Reply:
Thanks @kalafut! I didn't realize that `logical.ErrorResponse` accepted format strings. Addressed in https://github.com/hashicorp/vault/pull/10036/commits/50fbee14ca914c23f22170b0b86b2430142655af.